schnorr_fun/src/frost_backup.rs

```diff
@@ -0,0 +1,198 @@
+//!
+//! While this threshold will generally be ~small, in theory could be much higher than 32.
+//! Just doing up to 32 for now, but might be worth futureproofing also..
+//! We could also leave higher thresholds for a later backup version.
+//!
+//! ### Polynomial Identifier - (4)
+//! Hash of the polynomial coefficients and take the first 4 bech32 characters.
+//! This identifier will allow determination of secret share compatibility.
+//!
+//! It is possible to have two different polynomials that interpolate the same joint-secret,
+//! but shares on these different polynomials will not be compatible with one another.
+//!
+//! 32^4 = 2^20 ~> a collision rate of about one in a million.
+//!
+//! ### Secret Share - (52)
+//! A FROST secret share is fixed length scalar of 32 bytes.
+//! This is 32 * 8 / 5 = 51.2 -> 52 bech32 characters
+//!
+//! ### Share Index - (1 or 52)
+//! The share index can be be scalar, but will often be small for simplicity and smaller backups.
+//! By leaving this data piece at the end, we can use the length of the remaining data to
+//! easily decode a single bech32 char into integer, or 32 chars into a scalar.
+use alloc::{string::String, vec::Vec};
+use bech32::{self, u5, FromBase32, ToBase32, Variant::Bech32m};
+use secp256kfun::{
+    digest::{generic_array::typenum::U32, Digest},
+    marker::Secret,
+    Point, Scalar,
+};
+
+/// Secret shares can have a small index up to 32, or index at a specific scalar.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum BackupShareIndex {
+    /// a secret share at a small index up to 32
+    SmallIndex(usize),
+    /// a secret share at a specific scalar index
+    Scalar(Scalar),
+}
+
+/// An error encountered when encoding a Frostsnap backup.
+#[derive(Debug, Copy, Clone)]
+pub enum FrostBackupError {
+    /// Generic error from bech32 library
+    Bech32Error(bech32::Error),
+    /// Threshold was zero
+    ThresholdCantBeZero,
+    /// Threshold was greater than 32.
+    ThresholdTooBig,
+    /// Tried to decode a bech32 variant that was not bech32m
+    WrongBech32Variant(bech32::Variant),
+    /// Decoded secret share is not a valid secp256k1 scalar
+    InvalidSecretShareScalar,
+    /// Tried to decode backup with unknown prefix
+    InvalidHumanReadablePrefix,
+    /// The share index data length doesn't match expected for SmallIndex nor Scalar
+    UnknownShareIndexType,
+    /// Decoded share index is not a valid secp256k1 scalar
+    InvalidShareIndexScalar,
+}
+
+/// Create an identifier that's used to determine compatibility of FROST secret shares.
+/// The first 4 bech32 chars from a hash of the polynomial coefficients.
+/// Collision expected once in (32)^4 = 2^20.
+pub fn polynomial_identifier<H: Default + Digest<OutputSize = U32>>(
+    polynomial: Vec<Point>,
+) -> [u5; 4] {
+    let mut hash = H::default();
+    for coefficient in polynomial.into_iter() {
+        hash.update(coefficient.to_bytes());
+    }
+    hash.finalize().to_vec().to_base32()[0..4]
```

Comment:
The  `to_vec` seems unnecessary. Can't you call `to_base32` on the result of finalize?

I don't think this needs to be a separate function either?
I think you can do `hash.add(&polynomial[..])` if you import the `HashAdd` trait rather than do a loop.

Reply:
True it is now so simple it doesn't really need to be a separate function, but it might make things just a little bit easier for users of our FROST implementation, easily check their polynomial matches a backup using `polynomial_identifier()`.

& useful to separate for a spec